data/sea/54-zebra.h

```diff
@@ -241,23 +251,39 @@ zebra_state_t *zebra_alloc_state (piano_t *piano, zebra_config_t *cfg, int64_t a
     state->output_ptr   = output_ptr;
     state->output_fd    = output_fd;
 
+    state->drop_fd      = drop_fd;
+
     state->attribute_count    = attribute_count;
     state->entity_fact_offset = calloc(attribute_count, sizeof (int64_t));
     state->entity_fact_count  = calloc(attribute_count, sizeof (int64_t));
 
+    state->chunk_size         = cfg->chunk_size;
+    state->alloc_limit        = cfg->alloc_limit;
+    state->entity_alloc_count = 0;
+
     return state;
 }
 
 void zebra_collect_state (zebra_config_t *cfg, zebra_state_t *state)
 {
     cfg->entity_count = state->entity_count;
     cfg->fact_count = state->fact_count;
-    free (state->output_start);
     free (state->entity_fact_offset);
     free (state->entity_fact_count);
     free (state);
 }
 
+void zebra_write_dropped_entity (zebra_state_t *state, zebra_entity_t *entity) {
+    char *msg = calloc (error_msg_size, 1);
+    const size_t msg_size = snprintf (msg, error_msg_size, "%.*s\n", entity->id_length, entity->id_bytes);
+    write (state->drop_fd, msg, msg_size);
+    free (msg);
+}
+
+ibool_t zebra_limit_exceeded (zebra_state_t *state) {
+    return state->entity_alloc_count > state->alloc_limit;
```

Comment:
cool - I was just confused at first but makes sense. I don't see any better way to do it